Feature/intro (#122)
eProsima/Fast-RTPS-docs: docs/02-formalia/titlepage.rst

   :alt: eProsima
   :target: http://www.eprosima.com/

*eprosima Fast DDS* is a C++ implementation of the
`DDS (Data Distribution Service) Specification <https://www.omg.org/spec/DDS/About-DDS/>`__, a protocol
defined by the Object Management Group (OMG).
The *eprosima Fast DDS* library provides both an Application Programming Interface (API) and a communication protocol
that deploy
a Data-Centric Publisher-Subscriber (DCPS) model, with the purpose of establishing efficient and reliable
information distribution among Real-Time Systems.
*eprosima Fast DDS* is predictable, scalable, flexible, and efficient in resource handling.
For meeting these requirements, it makes use of typed interfaces and hinges on a many-to-many
distributed network paradigm that neatly allows separation of the publisher and subscriber sides of the communication.

> *eprosima Fast DDS* comprises:

Reviewer: This sentence could be the end of the previous paragraph

Author: which paragraph you mean?

Reviewer: Just the previous one, the one ending with "distributed network paradigm that neatly allows separation of the publisher and subscriber sides of the communication."

Author: You mean, not leaving the 18th line blank?

Reviewer: Yes!

> * The DDS API implementation.

Reviewer: Reference to DDS Layer section

Author: I think it's not too elegant to refer to the ToC here, and to the proper documentation in the two points below.

Reviewer: Sorry, I think I was not clear, I meant a reference to the documentation it self, specifically to :ref:`dds_layer`

> * A generation tool for bridging typed interfaces with the middleware implementation.

Reviewer: ```suggestion
* A :ref:`Fast DDS-Gen <fastrtpsgen_intro>` generation tool for bridging typed interfaces with the middleware
  implementation.
```

> * The underlying wire protocol implementation.

Reviewer: ```suggestion
* The underlying :ref:`RTPS <rtps_layer>` wire protocol implementation.
```

> .. figure:: /01-figures/DDS_concept_vertical.svg

Reviewer: ```suggestion
.. figure:: /01-figures/DDS_concept.svg
```

> .. figure:: /01-figures/DDS_concept_vertical.svg

Reviewer: I think the figure is out of context or explanation here. I'd move to right after the DDS API text, and maybe add a caption in the `figure` directive, as explained [here](http://openalea.gforge.inria.fr/doc/openalea/doc/_build/html/source/sphinx/rest_syntax.html?highlight=figure#figure-directive)

want to access portions of the data space declare their intent to become subscribers.
Each time a publisher posts new data into this space, the middleware propagates the information to all
interested subscribers.

The communication happens across domains, i. e. isolated abstract planes that link all the distributed applications
able to communicate with each other.
Only entities belonging to a same domain can interact, and the matching between entities subscribing to data and
entities publishing them is mediated by topics. Topics are unambiguous identifiers that associate a
name, which is unique in the domain, to a data type and a set of attached data-specific QoS.

DDS entities are modeled either as classes or typed interfaces.
The latter imply a more efficient resource handling as knowledge of the data
type prior to the execution allows allocating memory in advance rather than dynamically.

FastRTPSGen
^^^^^^^^^^^^

Relying on interfaces implies the need for a generation tool that translates type descriptions into appropriate
implementations that fill the gap between the interfaces and the middleware.

> This task is carried out by a dedicated generation tool, *FastRTPSGen*, a Java application that generates source code

Reviewer: In general
```suggestion
This task is carried out by a dedicated generation tool, *Fast DDS-Gen*, a Java application that generates source code
```

using the data types defined in an `Interactive Data Language (IDL) <https://www.omg.org/spec/IDL/About-IDL/>`__ file.

RTPS Wire Protocol
^^^^^^^^^^^^^^^^^^

The protocol used by *eProsima Fast DDS* to exchange messages over standard networks is the `Real-Time
Publish-Subscribe protocol (RTPS) <https://www.omg.org/spec/DDSI-RTPS/About-DDSI-RTPS/>`__, an interoperability wire
protocol for DDS defined and maintained by the OMG
consortium.
This protocol provides publisher-subscriber communications over transports such as TCP/UDP/IP, and guarantees
compatibility among different DDS implementations.

Given its publish-subscribe roots and its specification designed for meeting the same requirements addressed by the DDS

> application domain, the RTPS protocol is a natural choice for DDS implementations and maps naturally to many DDS

Reviewer: two naturals in a row

entities publishing them is mediated by topics. Topics are unambiguous identifiers that associate a
name, which is unique in the domain, to a data type and a set of attached data-specific QoS.

DDS entities are modeled either as classes or typed interfaces.
The latter imply a more efficient resource handling as knowledge of the data
type prior to the execution allows allocating memory in advance rather than dynamically.

.. figure:: /01-figures/DDS_concept.svg
    :align: center

    Conceptual diagram of how information flows within DDS domains.
    Only entities belonging to the same domain can discover each
    other through matching topics, and consequently exchange data between publishers and subscribers.

Fast DDS-Gen
^^^^^^^^^^^^

Relying on interfaces implies the need for a generation tool that translates type descriptions into appropriate
implementations that fill the gap between the interfaces and the middleware.
This task is carried out by a dedicated generation tool, :ref:`Fast DDS-Gen <fastrtpsgen_intro>`, a Java application
that generates source code using the data types defined in an

> `Interactive Data Language (IDL) <https://www.omg.org/spec/IDL/About-IDL/>`__ file.

Reviewer: ```suggestion
`Interface Definition Language (IDL) <https://www.omg.org/spec/IDL/About-IDL/>`__ file.
```

RTPS Wire Protocol
^^^^^^^^^^^^^^^^^^

The protocol used by *eProsima Fast DDS* to exchange messages over standard networks is the `Real-Time
Publish-Subscribe protocol (RTPS) <https://www.omg.org/spec/DDSI-RTPS/About-DDSI-RTPS/>`__, an interoperability wire
protocol for DDS defined and maintained by the OMG
consortium.
This protocol provides publisher-subscriber communications over transports such as TCP/UDP/IP, and guarantees
compatibility among different DDS implementations.

Given its publish-subscribe roots and its specification designed for meeting the same requirements addressed by the DDS
application domain, the RTPS protocol maps to many DDS concepts and is therefore a natural choice for DDS
implementations.
All the RTPS core entities are associated with an RTPS domain, which represents an isolated communication plane where
endpoints match.
The entities specified in the RTPS protocol are in one-to-one correspondence with the DDS entities, thus allowing
the communication to occur.

Main Features
^^^^^^^^^^^^^

> * **Two API Layers.** *eProsima Fast DDS* comprises a high-level publisher-subscriber layer focused on usability and a

Reviewer: ```suggestion
* **Two API Layers.** *eProsima Fast DDS* comprises a high-level DDS compliant layer focused on usability and a
```

>   lower-level RTPS layer that provides finer access to the RTPS protocol.

Reviewer: ```suggestion
  lower-level RTPS compliant layer that provides finer access to the RTPS protocol.
```

* **Real-Time behaviour.** *eProsima Fast DDS* can be configured to offer real-time features, guaranteeing responses
  within specified time constrains.

* **Built-in Discovery Service.** *eProsima Fast DDS* is based on the dynamical discovery of existing publishers and
  subscribers, and performs this task continuously without the need to contacting or setting any servers.
  However, a Client-Server discovery as well as other discovery paradigms can also be configured.

> * **Sync and Async publication modes.** *eProsima Fast DDS* supports both synchronous and asynchronous data access.

Reviewer: ```suggestion
* **Sync and Async publication modes.** *eProsima Fast DDS* supports both synchronous and asynchronous data publication.
```

  the network at any time without the
  need for reconfiguration.

* **Scalability and Flexibility.** DDS builds on the concept of a global data space. The middleware is in charge of
  propagating the information between publishers and subscribers. This guarantees that the distributed network is
  adaptable to reconfigurations and scalable to a large number of entities.

* **Application Portability.** The DDS specification includes a platform specific mapping to IDL, allowing an
  application using DDS to switch among DDS implementations with only a re-compile.

* **Extensibility.** *eProsima Fast DDS* allows the protocol to be extended and enhanced with new services without
  breaking backwards compatibility and interoperability.

* **Configurability and Modularity.** *eProsima Fast DDS* provides an intuitive way to be configured, either through
  code or XML profiles. Modularity allows simple devices to implement a subset of the protocol and still participate in
  the network.

* **High performance.** *eProsima Fast DDS* uses a static low-level serialization library,

>   `Fast CDR <https://github.com/eProsima/Fast-CDR>`,

Reviewer: ```suggestion
  `Fast CDR <https://github.com/eProsima/Fast-CDR>`__,
```

   :alt: eProsima
   :target: http://www.eprosima.com/

> *eprosima Fast DDS* is a C++ implementation of the DDS (Data Distribution Service) Specification, a protocol

Author: I agree but in Micro XRCE-DDS it's always eProsima in spite of eProsima's.
In Integration Service we followed the same convention.
In general I think it's desirable to follow the same syntax for all products.

> *eprosima Fast DDS* is a C++ implementation of the DDS (Data Distribution Service) Specification, a protocol

Author: ok.

> *eprosima Fast DDS* is a C++ implementation of the DDS (Data Distribution Service) Specification, a protocol

Reviewer: Sure

> *eprosima Fast DDS* is a C++ implementation of the DDS (Data Distribution Service) Specification, a protocol

Reviewer: Sorry, I did not mean that you should rename any file, I meant to change line 3 of index.rst so it says `eProsima Fast DDS Documentation`

> *eprosima Fast DDS* is a C++ implementation of the DDS (Data Distribution Service) Specification, a protocol

Author: ok. Got it.

> *eprosima Fast DDS* is a C++ implementation of the DDS (Data Distribution Service) Specification, a protocol

Author: It does say so right now. If you are ok with that, resolve the conversation.

defined by the Object Management Group (OMG).

> The *eprosima Fast DDS* library provides both an Application Interface (API) and a communication protocol that deploy

Author: ok.

a Data-Centric Publisher-Subscriber (DCPS) model, with the purpose of establishing efficient and reliable
information distribution among Real-Time Systems.

> *eprosima Fast DDS* is predictable, scalable, flexible, and efficient in resource handling.

Author: As above, I was following the convention used in Micro XRCE-DDS and Integration Service. I think we should align with Jaime (and maybe also Katrin?) regarding these issues, since it's important to have a common denominator for the company makeup.

> *eprosima Fast DDS* is predictable, scalable, flexible, and efficient in resource handling.

Reviewer: I just think it gets heavy to read over and over, but I'm ok to take it up to marketing

> *eprosima Fast DDS* is predictable, scalable, flexible, and efficient in resource handling.

Author: I totally agree. Let's discuss it with them maybe? For the moment I leave it this way.

For meeting these requirements, it makes use of typed interfaces and hinges on a many-to-many
distributed network paradigm that neatly allows separation of the publisher and subscriber sides of the communication.

*eprosima Fast DDS* comprises:

* The DDS API implementation.
* A generation tool for bridging typed interfaces with the middleware implementation.

> * The underlying wire protocol.

Author: ok.

> DDS API

Author: The title should be a link, you mean?

> DDS API

Reviewer: Yeah, that could work

> DDS API

Author: I didn't put the link in the end. Fist because there no concrete section to refer to, as we discussed, and second because I think that in the Intro you shouldn't be pointing directly to sections and stuff, because it's a still really primitive introduction to the documentation, one should read a little more before being redirected.

> ^^^^^^^

Author: I put it right above.

> The communication model adopted by DDS is one of a many-to-many unidirectional data exchange where the applications that

Author: ok.

> produce the data publish them to the local caches of subscribers belonging to applications that consume the data.

Author: ok. Even though 'data' can be considered either singular or plural.

> produce the data publish them to the local caches of subscribers belonging to applications that consume the data.

Reviewer: Data can be considered as the plural of datum for countable things. However, I think in this context the applications produce data as an unaccountable noun, thus my suggestion. I'm seeing more like a synonym of information here, where I think you are meaning the pieces of information. I'd stick with my suggestion but I see there are arguments on both sides. [data](https://www.oxfordlearnersdictionaries.com/definition/english/data?q=data)

> produce the data publish them to the local caches of subscribers belonging to applications that consume the data.

Author: changed to 'it'. Let's see how it looks.

The information flow is regulated by Quality of Service (QoS) policies established between the entities in
charge of the data exchange.

As a data-centric model, DDS builds on the concept of a "global data space" accessible to all interested applications.

> Applications that want to contribute information declare their intent to become publishers whereas applications that

Author: ok.

want to access portions of the data space declare their intent to become subscribers.
Each time a publisher posts new data into this space, the middleware propagates the information to all
interested subscribers.

> The communication happens across domains, isolated abstract planes that link all the distributed applications able to

Author: ok.

communicate with each other.
Only entities belonging to a same domain can interact, and the matching between entities subscribing to data and

> entities publishing them is mediated by topics. Topics are unambiguous identifiers that associate a

Author: same as above.

> entities publishing them is mediated by topics. Topics are unambiguous identifiers that associate a

Author: here I kept 'them'.

> name, that is unique in the domain, to a data type and a set of attached data-specific QoS.

Author: ok.

DDS entities are modeled either as classes or typed interfaces.
The latter imply a more efficient resource handling as knowledge of the data
type prior to the execution allows allocating memory in advance rather than dynamically.

> FastRTPSGen

Author: ok. Let's decide which will be the convention.

^^^^^^^^^^^^

Relying on interfaces implies the need for a generation tool that translates type descriptions into appropriate
implementations that fill the gap between the interfaces and the middleware.
This task is carried out by a dedicated generation tool, *FastRTPSGen*, a Java application that generates source code

> using the data types defined in an IDL file.

Author: ok. added the link as well.

RTPS Wire Protocol
^^^^^^^^^^^^^^^^^^

The protocol used by *eProsima Fast DDS* to exchange messages over standard networks is the Real-Time

> Publish-Subscribe protocol (RTPS), an interoperability wire protocol for DDS defined and maintained by the OMG

Author: ok.

consortium.
This protocol provides publisher-subscriber communications over transports such as TCP/UDP/IP, and guarantees
compatibility among different DDS implementations.

Given its publish-subscribe roots and its specification designed for meeting the same requirements addressed by the DDS
application domain, the RTPS protocol is a natural choice for DDS implementations and maps naturally to many DDS
concepts.
All the RTPS core entities are associated with an RTPS domain, which represents an isolated communication plane where
endpoints match.
The entities specified in the RTPS protocol are in one-to-one correspondence with the DDS entities, thus allowing
the communication to occur.

> Main Features

Author: Same as above, I was following the Micro XRCE-DDS convention. But I can agree.

> Main Features

Author: Ok, I'll start working on it.

> Main Features

Reviewer: I'd say let's do it. The list like this is ugly and difficult to follow

> Main Features

Author: Agreed.

> Main Features

Reviewer: I actually like the long list, although I think the order is somewhat estrange

> Main Features

Author: Yes, the order is totally random.

> Main Features

Reviewer: I'd try to come up with something, probably what sells best from a commercial point of view

> Main Features

Author: ok. I try to reorganize it, according to some arbitrary criterium.

^^^^^^^^^^^^^

**High performance.**

> *eProsima Fast DDS* uses a static low-level serialization library (*eProsima Fast CDR*) that serializes in *CDR*.

Author: ok.

> *eProsima Fast DDS* uses a static low-level serialization library (*eProsima Fast CDR*) that serializes in *CDR*.

Reviewer: Maybe mention that it is defined in [here](https://www.omg.org/spec/DDSI-RTPS/) (chapter about data encapsulation)

> *eProsima Fast DDS* uses a static low-level serialization library (*eProsima Fast CDR*) that serializes in *CDR*.

Author: ok.

**Low resources consumption.**

*eProsima Fast DDS*:

* Allows to pre-allocate resources, to minimize dynamic resource allocation.
* Avoids the use of unbounded resources.
* Minimizes the need to copy data.

**Multi-platform.**

The OS dependencies are treated as pluggable modules.
The user can easily implement his platform modules to *eProsima Fast DDS* library in his specific platform.

> By default, the project can run over Linux and Windows.

Author: ok.

**Free and Open Source.**

The Fast DDS library, the underneath RTPS library, the generator tool, the internal dependencies (such as *eProsima
Fast CDR*) and the external ones (such as the *foonathan* library) are free and open source.

**Easy to use.**

The project comes with an out-of-the-box example, the *DDSHelloWorld* (TODO: add ref) that puts into communication a
publisher and a subscriber, showcasing how *eProsima Fast DDS* is deployed.
Additionally, the interactive demo *ShapesDemo* is available for the user to dive into the DDS world.

> Both the DDS and the RTPS layers are thoroughly explained in the Fast DDS section (TODO: add ref).

Author: ok.

**Best effort and reliable communication.**

*eProsima Fast-DDS* supports both *best effort*, for fast and light communication, and *reliable*, when reliability

> is needed.

Author: Changed it as:
*eProsima Fast-DDS* supports an optional reliable communication paradigm over *Best Effort* communications protocols such as UDP. Furthermore, another way of setting a reliable communication is to use our TCP transport.

> **Built-in Discovery Service.**

Author: 1) Link in the title?
2) ok.

> **Built-in Discovery Service.**

Author: I think that at this point either we put references everywhere, or nowhere. I'd vote for the nowhere option, see my commentary above.

> *eProsima Fast DDS* is based on the dynamical discover of the existence of publishers and subscribers

Author: ok.

and performs this task continuously without

> the need to contact any servers.

Author: 'or set'
to maintain the verbal conjugation.

> the need to contact any servers.

Reviewer: Then I'd go with gerund for both

> the need to contact any servers.

Author: ok. I think it's present continuous, though.

New applications and services are automatically discovered, and can join and leave the network at any time without the
need for reconfiguration.

**Scalability and Flexibility**

DDS builds on the concept of a global data space. The middleware is in charge of propagating the information between
publishers and subscribers. This guarantees that the distributed network is adaptable to reconfigurations
and scalable to a large number of entities.

**Application Portability**

The DDS specification includes a platform specific mapping to IDL, allowing an application using
DDS to switch among DDS implementations with only a re-compile.

**Extensibility**

*eProsima Fast DDS* allows the protocol to be extended and enhanced with new services without
breaking backwards compatibility and interoperability.

**Configurability and Modularity**

*eProsima Fast DDS* provides an intuitive way to be configured, either through code or XML profiles.
Modularity allows simple devices to implement a subset of the protocol and still participate in
the network.

**Two API Layers**

*eProsima Fast DDS* comprises a high-level publisher-subscriber layer focused on usability and a lower-level 

> writer-reader layer that provides finer access to the RTPS protocol.

Author: ok.

**Commercial support**

> Available at support@eprosima.com

Author: like phone number/link to the company webpage?

> Available at support@eprosima.com

Reviewer: More like what can they ask. Something more like "do you need this or that just write to us" kind of way

> Available at support@eprosima.com

Author: I didn't change much, but added some more contacts. If you want it more 'interactive' and 'informal', I can make it.

> Available at support@eprosima.com

Reviewer: Fine by me

Structure of the Documentation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

> This documentation is organized into the following sections:

Reviewer: I think this section looks weird in the HTML, maybe the list in the end with a description would suffice

Author: Ok, so deleting lines 153-175, and maybe just leave the sentence:
'This documentation is organized into the following sections:'
?

Reviewer: Yep, that's what I thought
